### PR TITLE
transport-cc parsing to be compliant with libwebrtc, make transport-cc reference time to be unsigned, therefore fix 357.

### PR DESCRIPTION
**Pull request by sarumjanuch**
Changes and tests to make sure mediasoup is parsing transport-cc RTCP feedback packages the same way as libwebrtc does. 
Metadata in commit was created by parsing buffers in libwebrtc project.
Also this PR includes changes https://github.com/versatica/mediasoup/pull/900 made by @penguinol. 
This PR makes sure that there could be no negative reference time (event though https://datatracker.ietf.org/doc/html/draft-holmer-rmcat-transport-wide-cc-extensions-01#section-3.1 mandates that reference time is 24 bit signed integer, libwebrtc is ignoring that, and with with PR we too), and also handles wraparound, to be consistent with latest libwebrtc code. Therefore it rolls back mediasoup changes in libwebrtc related to https://github.com/versatica/mediasoup/issues/357, and https://github.com/versatica/mediasoup/issues/357 should never happen again.

### CODE REVIEW

File: worker/test/src/RTC/RTCP/TestFeedbackRtpTransport.cpp [hunk 1]
```diff
@@ -571,4 +571,223 @@ SCENARIO("RTCP Feeback RTP transport", "[parser][rtcp][feedback-rtp][transport]"
...
+                        .buffer = std::vector<uint8_t> {
+                                0xAF, 0xCD, 0x00, 0x06, 0xFA, 0x17, 0xFA, 0x17,
+                                0x08, 0xEB, 0x06, 0xD7, 0x00, 0x13, 0x00, 0x04,
+                                0x0C, 0x89, 0x16, 0x02, 0x20, 0x04, 0x10, 0x00,
+                                0x10, 0x00, 0x00, 0x02
+                        }
+                }
+        };
+
+        for (const auto &packetMeta: feedbackPacketsMeta) {
```
Comment (ibc) [inline worker/test/src/RTC/RTCP/TestFeedbackRtpTransport.cpp]: Please run `make lint` in worker/ directory. It should complain. `make format` will fix it.

**Review by ibc — CHANGES_REQUESTED** [review]

File: worker/test/src/RTC/RTCP/TestFeedbackRtpTransport.cpp [hunk 2]
```diff
@@ -572,222 +572,163 @@ SCENARIO("RTCP Feeback RTP transport", "[parser][rtcp][feedback-rtp][transport]"
...
+			uint32_t baseTimeRaw;
+			uint64_t baseTimeMs;
+			uint16_t baseSequence;
+			size_t packetStatusCount;
+			std::vector<int16_t> deltas;
+			std::vector<uint8_t> buffer;
+		} FeedbackPacketsMeta;
+
+		// Metadata collected by parsing buffers with libwebrtc, buffers itself.
+		// were generated by chrome in direction of mediasoup.
```
Comment (ibc) [inline worker/test/src/RTC/RTCP/TestFeedbackRtpTransport.cpp:591]:
```suggestion
		// Were generated by chrome in direction of mediasoup.
```

**Review by ibc — APPROVED** [review]

**Comment by ibc** [comment]
There are still lint errors: https://github.com/versatica/mediasoup/runs/8097882496?check_suite_focus=true

`make format` should fix them.

**Comment by ibc** [comment]
Note that you can use `// clang-format off` and `on` to delimit C++ blocks for which lint should not be checked. Check `RequestChannel.cpp` file if you find it useful within this PR.

**Review by jmillan — APPROVED** [review]

File: worker/include/RTC/RTCP/FeedbackRtpTransport.hpp [hunk 3]
```diff
@@ -42,6 +42,8 @@ namespace RTC
 		class FeedbackRtpTransportPacket : public FeedbackRtpPacket
 		{
 		public:
+			static constexpr int64_t kBaseTimeTick   = 64;
+			static constexpr int64_t kTimeWrapPeriod = kBaseTimeTick * (1ll << 24);
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackRtpTransport.hpp]: Blank line after this, please?

File: worker/include/RTC/RTCP/FeedbackRtpTransport.hpp [hunk 4]
```diff
@@ -237,9 +239,29 @@ namespace RTC
 			{
 				return this->referenceTime;
 			}
+			// We only use this for testing purpose
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackRtpTransport.hpp]: Dot at the end of the comment, please 😊

File: worker/include/RTC/RTCP/FeedbackRtpTransport.hpp [hunk 5]
```diff
@@ -290,7 +312,7 @@ namespace RTC
 
 		private:
 			uint16_t baseSequenceNumber{ 0u };
-			int32_t referenceTime{ 0 };
+			uint32_t referenceTime{ 0 };
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackRtpTransport.hpp:317]: Are we sure of this? Reference time is supposed to be a 24bits signed value.

File: worker/include/RTC/RTCP/FeedbackRtpTransport.hpp (hunk 3 above)
Comment (jmillan) [reply worker/include/RTC/RTCP/FeedbackRtpTransport.hpp]: Also please, remove the "k" prefix from constant name as it's not mediasoup style but libwebrtc.

File: worker/include/RTC/RTCP/FeedbackRtpTransport.hpp (hunk 5 above)
Comment (sarumjanuch) [reply worker/include/RTC/RTCP/FeedbackRtpTransport.hpp:317]: Yes, the whole point of this commit is to not follow that rule, as libwebrtc doesn't care about sign for a reference_time already. https://webrtc.googlesource.com/src/+/refs/heads/main/modules/rtp_rtcp/source/rtcp_packet/transport_feedback.h#170

File: worker/include/RTC/RTCP/FeedbackRtpTransport.hpp (hunk 3 above)
Comment (ibc) [reply worker/include/RTC/RTCP/FeedbackRtpTransport.hpp]: And should it start with upper letter? I cannot remember.

Comment (sarumjanuch) [reply worker/include/RTC/RTCP/FeedbackRtpTransport.hpp]: Made without k, and starting with a capital letter.

**Review by jmillan — APPROVED** [review]

File: worker/include/RTC/RTCP/FeedbackRtpTransport.hpp [hunk 6]
```diff
@@ -237,9 +241,29 @@ namespace RTC
...
+			}
 			int64_t GetReferenceTimestamp() const // Reference time in ms.
 			{
-				return static_cast<int64_t>(this->referenceTime) * 64;
+				return TimeWrapPeriod + static_cast<int64_t>(this->referenceTime) * BaseTimeTick;
+			}
+			int64_t GetBaseDelta(const int64_t previousTimestampMs) const
+			{
+				int64_t delta = GetReferenceTimestamp() - previousTimestampMs;
+				// Compensate for wrap around.
```
Comment (ibc) [inline worker/include/RTC/RTCP/FeedbackRtpTransport.hpp:256]: Add a blank line before the comment, please.

**Review by ibc — APPROVED** [review]
Nice. Just a cosmetic change requested. Also, if this PR fixes an existing issue in GH (I think there is a reported crash, am I right?) reference it in the PR description so it will be automatically closed. Tell me when those things are done and let's merge then.

**Comment by sarumjanuch** [comment]
> Nice. Just a cosmetic change requested. Also, if this PR fixes an existing issue in GH (I think there is a reported crash, am I right?) reference it in the PR description so it will be automatically closed. Tell me when those things are done and let's merge then.
@ibc Seems like done, please run tests, because I am not sure that clang formatter is not playing games with me again.

**Comment by ibc** [comment]
Running!

**Comment by ibc** [comment]
Please re-request review once tests pass.

**Review by ibc — APPROVED** [review]

**Comment by ibc** [comment]
> Nice. Just a cosmetic change requested. Also, if this PR fixes an existing issue in GH (I think there is a reported crash, am I right?) reference it in the PR description so it will be automatically closed. Tell me when those things are done and let's merge then.

Will merge. Can you check this please? I think there is an open issue or PR (here in GH) that aims to fix an issue that may be fixed by this PR, but maybe I'm wrong.

**Comment by ibc** [comment]
Oh, I request a change: please add a item in Changelog.md following the style of other items.

**Comment by sarumjanuch** [comment]
> Will merge. Can you check this please? I think there is an open issue or PR (here in GH) that aims to fix an issue that may be fixed by this PR, but maybe I'm wrong.

Yes, i did that, also added in title.

**Comment by sarumjanuch** [comment]
> Oh, I request a change: please add a item in Changelog.md following the style of other items.

doing right now

**Pull request merged**